app-backend/db/src/main/scala/TaskDao.scala

```diff
@@ -294,4 +294,111 @@ object TaskDao extends Dao[Task] {
   def deleteLayerTasks(projectId: UUID, layerId: UUID): ConnectionIO[Int] = {
     (fr"DELETE FROM " ++ this.tableF ++ fr"WHERE project_id = ${projectId} and project_layer_id = ${layerId}").update.run
   }
+
+  def getTaskActionTimeF(projectId: UUID,
+                         layerId: UUID,
+                         fromStatus: TaskStatus,
+                         toStatus: TaskStatus,
+                         params: UserTaskActivityParameters): Fragment = {
+    val selectTaskActionF =
+      fr"SELECT task_actions.user_id, tasks.id as task_id, MAX(task_actions.timestamp) AS timestamp FROM"
+    val joinTaskActionF =
+      fr"tasks LEFT JOIN task_actions ON task_actions.task_id = tasks.id"
+    val whereClause = Fragments
+      .whereAndOpt(
+        Some(fr"project_id = uuid($projectId)"),
+        Some(fr"project_layer_id = uuid($layerId)"),
+        Some(fr"task_actions.FROM_status = $fromStatus"),
+        Some(fr"task_actions.to_status = $toStatus"),
+        params.actionStartTime map { start =>
+          fr"task_actions.timestamp >= $start"
+        },
+        params.actionEndTime map { end =>
+          fr"task_actions.timestamp <= $end"
+        },
+        params.actionUser map { userId =>
+          fr"task_actions.user_id = $userId"
+        }
+      )
+    val groupByF =
+      fr"GROUP BY (task_actions.user_id, tasks.id, task_actions.FROM_status, task_actions.to_status)"
+    fr"(" ++ selectTaskActionF ++ joinTaskActionF ++ whereClause ++ groupByF ++ fr")"
+  }
+
+  def getTaskCountAndTimeSpentByActionF(
+      projectId: UUID,
+      layerId: UUID,
+      action: String,
+      params: UserTaskActivityParameters): Fragment = {
+    val taskSelectF = fr"""
+      SELECT
+        a.user_id,
+        count(a.task_id) AS task_count,
+        sum(EXTRACT(EPOCH FROM (b.timestamp - a.timestamp))) / count(a.task_id) AS avg_time
+      FROM
+    """
```

Comment:
We'll want to keep an eye on the performance of this query. Depending on the time period / number of entries, this could easily balloon.

We'll probably want to put limitations on the length of time it can cover

Reply:
Yes, I agree that we need to keep an eye on the performance of this query.

In terms of the length of time, we have these filters on annotate frontend: last week, last month, last year, all time

Reply:
We might be able to get away with just getting rid of the "all time" option. Query time will scale linearly with the number of events, so we've got some time to figure out it out.

Reply:
One good thing is that we can make some reasonable assumptions about the max number of tasks (something like 50k) and an average number of actions per task and get a good idea of what the upper bound might be.